### DISCUSSION TITLE
No module named 'pandas'

### DISCUSSION THREAD

**Discussion by abdokhalil07** [flet-dev/flet#4512] [Q&A] [answered]
### Question

i use python 3.13 flet 0.25.1
after build exe i get errore"No module named 'pandas'"

### Error message

```bash
Traceback (most recent call last):
  File "<string>", line 47, in <module>
  File "<frozen runpy>", line 229, in run_module
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\abdelrahman\AppData\Roaming\Flet\attendance\flet\app\main.py", line 2, in <module>
    import pandas as pd
ModuleNotFoundError: No module named 'pandas'
```

### ------------------------------------------------------

- [X] I have searched for answers to my question both in the issues and in previous discussions.

**Comment by ndonkoHenri** [comment, marked as answer]
Can you share your requirements or pyproject file?

**Reply by abdokhalil07** [reply]
i forget to add pandas to pyproject 
thanks

**Reply by abdokhalil07** [reply]
i have a problem with app after build it to exe 
i select csv file and make some process and convert it to xlsx
it work in vscode after build exe not work
My code

`
import flet as ft
import pandas as pd
import calendar
from datetime import datetime, timedelta
import csv
import os

def main(page: ft.Page):

    lst =[]
    def on_dialog_result(e: ft.FilePickerResultEvent):
        try:
            lst.clear()
            lst.append(e.files[0].path)
            file_name_text.value = e.files[0].path
            download_btn.visible = True
            page.update()
            return e.files[0].path
        except: pass

    # [+] Convert CSV File To JSON File ---------------------------------
    def csv_to_json(file_path):
        data = []
        month_lst = []
        code_lst = []
        with open(file_path, 'r') as csv_file:
            csv_reader = csv.DictReader(csv_file)
            for i in csv_reader:
                time1_str = str(float(i['out']))
                time2_str = str(float(i['in']))

               
                time_format = "%H.%M"
                time1 = datetime.strptime(time1_str, time_format)
                time2 = datetime.strptime(time2_str, time_format)

                
                time2_delta = timedelta(hours=time2.hour, minutes=time2.minute)
                
                if time1 < time2:
                    time1 += timedelta(days=1)
                time_difference = time1 - time2_delta
                code_lst.append(i['code'])
                month_lst.append(f"{datetime.strptime(i['date'],"%d/%m/%Y").month}-{datetime.strptime(i['date'],"%d/%m/%Y").year}")
                i['total'] = f'{time_difference.hour:02d}:{time_difference.minute:02d}'
                i['month'] = f"{datetime.strptime(i['date'],"%d/%m/%Y").month}-{datetime.strptime(i['date'],"%d/%m/%Y").year}"
                data.append(i)
        code_lst = list(set(code_lst))
        month_lst = list(set(month_lst))

        mounth_log = []

        for mon in month_lst:
            month_dic = {'name':mon, 'data':[]}
            for i in code_lst :
                d_ = {'code': i, "log":[]}
                for x in data:
                    if mon == x['month']:
                        if i == x['code']:
                            d_['log'].append({'date':x['date'] ,'in': x['in'] ,'out': x['out'] ,'total': x['total']})
                month_dic['data'].append(d_)
            mounth_log.append(month_dic)

        atednance = {'mounths':mounth_log}
        return atednance
    # [+] Export EXCEL SUMMARY WORKBOOK
    def generate_month_days(year, month):
        days_in_month = calendar.monthrange(year, month)[1]
        return [f"{day:02d}/{month:02d}/{year}" for day in range(1, days_in_month + 1)]

    
    def convert_to_minutes(time_str):
        hours, minutes = map(int, time_str.split(':'))
        return hours * 60 + minutes

    # Main function 
    def process_attendance(file_path, output_file_path):

        json_data = csv_to_json(file_path)
        
        expanded_records = []
        v_days_summary = []

        for month in json_data['mounths']:
            month_name = month['name']
            year, month_num = map(int, month_name.split('-')[::-1]) 
            all_days = generate_month_days(year, month_num)
            total_days_in_month = len(all_days)
            for entry in month['data']:
                code = entry['code']
                existing_dates = {log['date']: log for log in entry['log']}
                total_minutes = 0
                v_days_count = 0
                actual_days_count = 0
                for day in all_days:
                    if day in existing_dates:
                        log = existing_dates[day]
                        expanded_records.append({
                            'Month': month_name,
                            'Code': code,
                            'Date': day,
                            'In': log['in'],
                            'Out': log['out'],
                            'Total': log['total']
                        })
                        total_minutes += convert_to_minutes(log['total'])
                        actual_days_count += 1
                    else:
                        expanded_records.append({
                            'Month': month_name,
                            'Code': code,
                            'Date': day,
                            'In': 'V',
                            'Out': 'V',
                            'Total': 'V'
                        })
                        v_days_count += 1
                
                
                total_hours = total_minutes // 60
                total_remaining_minutes = total_minutes % 60
                v_days_summary.append({
                    'Month': month_name,
                    'Code': code,
                    'Total Days in Month': total_days_in_month,
                    'Actual Recorded Days': actual_days_count,
                    'Days with V': v_days_count,
                    'Total Hours': f"{total_hours}:{total_remaining_minutes:02d}",
                })

        
        expanded_df = pd.DataFrame(expanded_records)
        v_summary_df = pd.DataFrame(v_days_summary)

        
        with pd.ExcelWriter(output_file_path, engine='openpyxl') as writer:
            expanded_df.to_excel(writer, sheet_name='Expanded Logs', index=False)
            v_summary_df.to_excel(writer, sheet_name='Monthly Summary', index=False)
        print("[+] DONE")

    # [+] Download EXCELL File
    def download_file(e):
        user = os.getlogin()
        x= f'C:\\Users\\{user}\\Downloads'
        output_file_path = f'{x}\\attendance_final.xlsx'  
        process_attendance(lst[0], output_file_path)
        lst.clear()
        download_btn.visible = False
        file_name_text.value
        page.update()

    file_picker = ft.FilePicker(on_result=on_dialog_result)
    page.overlay.append(file_picker)
    download_btn = ft.ElevatedButton('Download', on_click=download_file, visible=False)
    file_name_text = ft.Text()
    page.add(
        ft.Column([ft.ElevatedButton("Choose files...",
            on_click=lambda _: file_picker.pick_files(allow_multiple=False, allowed_extensions=["csv"])), file_name_text]),
        download_btn
    )
    page.window.width = 500
    page.window.height = 500
    page.update()
ft.app(main)

`

**Reply by ndonkoHenri** [reply]
What exactly doesn't work? Can you share a video recording?

**Reply by abdokhalil07** [reply]
https://github.com/user-attachments/assets/814d3ef9-c430-4dcc-8e05-fcb6d5ecd065

**Reply by FeodorFitsner** [reply]
Can you also share a full build log please with '-v' option?

**Reply by abdokhalil07** [reply]
(venv) PS D:\ABDO\python_flet\projects\attendance> flet build -v windows
[21:45:13] No additional Flutter dependencies!
[21:45:18] Created Flutter bootstrap project from gh:flet-dev/flet-build-template with ref 0.25.1 ✅                                                                                                    
           Copying D:\ABDO\python_flet\projects\attendance\src\assets\icon.png to D:\ABDO\python_flet\projects\attendance\build\flutter\images
           Customized app icons and splash images ✅                                                                                                                                                    
           Run subprocess: ['C:\\flutter\\bin\\dart.BAT', 'run', 'flutter_launcher_icons']
(     ●) Generating app icons(4.9s) 
( ●    ) Generating app icons ⏳...(1.1s)
( ●    ) Generating app icons ⏳...Got dependencies in `D:\ABDO\python_flet\projects\attendance\build\flutter`.
(  ●   ) Generating app icons(9.7s) 
Built flutter_launcher_icons:flutter_launcher_icons.
(   ●  ) Generating app icons ⏳...   ════════════════════════════════════════════
     FLUTTER LAUNCHER ICONS (v0.14.2)
  ════════════════════════════════════════════
  
(●     ) Generating app icons ⏳...• Creating default icons Android
( ●    ) Generating app icons ⏳...• Overwriting the default Android launcher icon with a new icon
(   ●  ) Generating app icons ⏳...• Creating adaptive icons Android
(     ●) Generating app icons ⏳...• No colors.xml file found in your Android project
• Creating colors.xml file and adding it to your Android project
• Creating mipmap xml file Android
(  ●   ) Generating app icons ⏳...• Overwriting default iOS launcher icon with new icon
(   ●  ) Generating app icons ⏳...Creating Icons for Web...
(  ●   ) Generating app icons ⏳...Creating Icons for Windows...
(   ●  ) Generating app icons ⏳...Creating Icons for MacOS...
( ●    ) Generating app icons ⏳... 
✓ Successfully generated launcher icons
[21:45:38] Generated app icons ✅                                                                                                                                                                       
           Run subprocess: ['C:\\flutter\\bin\\dart.BAT', 'run', 'serious_python:main', 'package', 'D:\\ABDO\\python_flet\\projects\\attendance\\src', '--platform', 'Windows', '--requirements',       
           'flet==0.25.1,pandas', '--exclude', 'build', '--cleanup']
(●     ) Packaging Python app ⏳... 
Built serious_python:main.
(    ● ) Packaging Python app ⏳...Running package command
Extra PyPi indexes: [https://pypi.flet.dev]
Creating asset directory: D:\ABDO\python_flet\projects\attendance\build\flutter\app     
Created temp directory: C:\Users\ABDELR~1\AppData\Local\Temp\serious_python_temp80be81e1
Copying Python app from D:\ABDO\python_flet\projects\attendance\src to a temp directory 
Cleanup app
Configured Windows/ platform with sitecustomize.py
Installing [flet==0.25.1, pandas] with pip command to C:\Users\ABDELR~1\AppData\Local\Temp\serious_python_temp80be81e1\__pypackages__
(   ●  ) Packaging Python app ⏳... Downloading Python distributive from https://github.com/indygreg/python-build-standalone/releases/download/20240909/cpython-3.12.6+20240909-x86_64-pc-windows-msvc-sh
ared-install_only_stripped.tar.gz to a build directory
( ●    ) Packaging Python app ⏳...Extracting Python distributive
( ●    ) Packaging Python app ⏳...Cleanup installed packages
(   ●  ) Packaging Python app ⏳...Creating app archive at D:\ABDO\python_flet\projects\attendance\build\flutter\app/app.zip from a temp directory
(    ● ) Packaging Python app ⏳...Writing app archive hash to D:\ABDO\python_flet\projects\attendance\build\flutter\app/app.zip.hash
(  ●   ) Packaging Python app ⏳...Deleting temp directory
[21:47:11] Packaged Python app ✅                                                                                                                                                                       
           Run subprocess: ['C:\\flutter\\bin\\flutter.BAT', 'build', 'windows', '--build-name', '0.1.0']
(  ●   ) Building Windows app ⏳... 
(   ●  ) Building Windows app ⏳... Downloading packages...
(●     ) Building Windows app ⏳...  archive 3.6.1 (4.0.0 available)
  characters 1.3.0 (1.3.1 available)
  clock 1.1.1 (1.1.2 available)
  collection 1.18.0 (1.19.1 available)
  flutter_lints 2.0.3 (5.0.0 available)
  http_parser 4.0.2 (4.1.1 available)
  intl 0.19.0 (0.20.1 available)
  js 0.6.7 (0.7.1 available)
  lints 2.1.1 (5.1.0 available)
  material_color_utilities 0.11.1 (0.12.0 available)
  meta 1.15.0 (1.16.0 available)
  path 1.9.0 (1.9.1 available)
  sensors_plus 4.0.2 (6.1.0 available)
  sensors_plus_platform_interface 1.2.0 (2.0.0 available)
  serious_python 0.8.3 (0.8.4 available)
  shelf 1.4.1 (1.4.2 available)
  toml 0.15.0 (0.16.0 available)
  url_strategy 0.2.0 (discontinued)
! wakelock_plus 1.2.8 (overridden)
! web 1.1.0 (overridden)
  web_socket_channel 2.4.5 (3.0.1 available)
! window_manager 0.4.3 (overridden)
Got dependencies!
1 package is discontinued.
19 packages have newer versions incompatible with dependency constraints.
Try `flutter pub outdated` for more information.
(   ●  ) Building Windows app ⏳... 
(    ● ) Building Windows app ⏳...windows/flutter/CMakeLists.txt does not use FLUTTER_TARGET_PLATFORM, updating.
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.41.34120\include\format(42): warning STL4038: The contents of <format> are available only with C++20 or later. [D:\ABDO\python_flet\projects\attendance\build\flutter\build\windows\x64\plugins\serious_python_windows\serious_python_windows_plugin.vcxproj]
(  ●   ) Building Windows a  172.6s 
√ Built build\windows\x64\runner\Release\attendance.exe
[21:50:07] Built Windows app ✅                                                                                                                                                                         
           Copying build output from: D:\ABDO\python_flet\projects\attendance\build\flutter\build\windows\x64\runner\Release\*
[21:50:08] Copied build to build\windows directory ✅                                                                                                                                                   
           Successfully built your Windows app! 🥳 Find it in  build\windows directory. 📁

**Reply by FeodorFitsner** [reply]
What Python dependencies does the app have?

**Reply by abdokhalil07** [reply]
Python 3.13.0

**Reply by FeodorFitsner** [reply]
I mean what's in requirements.txt?

**Reply by abdokhalil07** [reply]
"flet==0.25.1",
  "pandas"

**Reply by abdokhalil07** [reply]
i found the problem ,
 pandas depends on "openpyxl" to handle xlsx files, 
 i should add "openpyxl" to my requirements.
 i try it and work
thanks for help
